### ISSUE TITLE
what's the difference between your BCNN and BCNN with dropout in testing?

### ISSUE THREAD

**Issue by ShellingFord221** [kumar-shridhar/PyTorch-BayesianCNN#11]
Hi,  I've been following your work for a few months. To put it simply, your model randomly samples weight w to make traditional CNN into a Bayesian way, where w is drawn from a Gaussian distribution (e.g. line 148 in BBBlayers.py: weight = self.conv_qw_mean + sigma_weight * self.eps_weight.normal_()). But in another BCNN model (Deep Bayesian Active Learning, https://arxiv.org/pdf/1703.02910.pdf), the author thinks that Bayesian characteristic reflects in using dropout when testing the model (see the issue that I discussed with damienlancry: https://github.com/Riashat/Deep-Bayesian-Active-Learning/issues/5) and other modules remain unchanged (e.g., the layers and the weights).

I'm really confused that what on earth does Bayes mean in a neural network like CNN. I'm very looking forward to your reply!

**Comment by kumar-shridhar** [comment]
Hi. There are many differences actually. In Gal and Ghahramani paper, they use Dropout as a way to induce Bayesian characteristics. As you mentioned, using it during test time. In our case, we say every weight in the network is a distribution and we approximate it using Variational Inference and use local reparameterization ( self.conv_qw_mean + sigma_weight * self.eps_weight.normal_())  to train it using a gradient-based learning method like backprop. 

Bayesian can induce many characteristics in a network that normal network lacks: regularization effect to prevent overfitting (like Dropout), less overconfident prediction and sampling and adding Gaussian Noise makes it more robust. Also, we can estimate uncertainties in a network.
